### PR TITLE
Fixes for HoloLens builds including Vuforia and OpenXR update closing #1170

### PR DESCRIPTION
**Pull request by robhillman97**
Includes;
- Update to use OpenXR
- Update to Vuforia 10.15.4
- Addition of world space console window
- Multiple fixes for various HoloLens issues detailed [here](https://github.com/WEKIT-ECS/MIRAGE-XR/milestone/17)

### CODE REVIEW

File: Assets/MirageXR/Common/Scripts/CombinedEditor/ActionDetailView/Annotation Editors/ImageMarker Prefabs/ImageMarkerController.cs [hunk 1]
```diff
@@ -12,6 +13,7 @@ public class ImageMarkerController : MirageXRPrefab
 
         private string _imageName;
         private ToggleObject _content;
+        public IImageTarget target;
```
Comment (cawke) [inline Assets/MirageXR/Common/Scripts/CombinedEditor/ActionDetailView/Annotation Editors/ImageMarker Prefabs/ImageMarkerController.cs]:
```suggestion
        public IImageTarget _target;
```

File: Assets/MirageXR/Common/Scripts/Managers/ImageTarget/Manager/ImageTargetManagerVuforia.cs [hunk 2]
```diff
@@ -9,19 +8,14 @@ public class ImageTargetManagerVuforia : ImageTargetManagerBase
...
-        if (vuforiaRuntime.InitializationState == VuforiaRuntime.InitState.NOT_INITIALIZED)
-        {
-            vuforiaRuntime.InitVuforia();
-        }
-
-        while (vuforiaRuntime.InitializationState != VuforiaRuntime.InitState.INITIALIZED)
+        while (!vuforiaRuntime.IsInitialized)
         {
             await Task.Yield();
         }
```
Comment (cawke) [inline Assets/MirageXR/Common/Scripts/Managers/ImageTarget/Manager/ImageTargetManagerVuforia.cs]: activate **delayed initialization** and call the initialization here.

image

```
        var vuforiaRuntime = VuforiaApplication.Instance;

        if (!vuforiaRuntime.IsInitialized)
        {
            vuforiaRuntime.Initialize();
            while (!vuforiaRuntime.IsInitialized)
            {
                await Task.Yield();
            }
        }
```

File: Assets/MirageXR/Common/Scripts/Managers/ImageTarget/Target/ImageTargetVuforia.cs [hunk 3]
```diff
@@ -9,27 +9,36 @@ public class ImageTargetVuforia : ImageTargetBase
 
     public ImageTargetBehaviour imageTargetBehaviour => _imageTargetBehaviour;
 
+    private Status _oldTargetStatus;
```
Comment (cawke) [inline Assets/MirageXR/Common/Scripts/Managers/ImageTarget/Target/ImageTargetVuforia.cs:12]:
```
public class ImageTargetVuforia : ImageTargetBase
{
    private ImageTargetBehaviour _imageTargetBehaviour;
    private Status _oldTargetStatus;

    public ImageTargetBehaviour imageTargetBehaviour => _imageTargetBehaviour;

    protected override void TrackerInitialization()
    {
        _imageTargetBehaviour = GetComponent<ImageTargetBehaviour>();
        _oldTargetStatus = _imageTargetBehaviour.TargetStatus.Status;
        _imageTargetBehaviour.OnTargetStatusChanged += OnTrackableStatusChangedAction;
    }

    private void OnTrackableStatusChangedAction(ObserverBehaviour observerBehaviour, TargetStatus targetStatusChangeResult)
    {
        var statusChangeResult = targetStatusChangeResult.Status;
        OnStateChanged(ToTrackingState(_oldTargetStatus), ToTrackingState(statusChangeResult));
        _oldTargetStatus = statusChangeResult;
    }

    private static TrackingState ToTrackingState(Status state)
    {
        return state switch
        {
            Status.NO_POSE => TrackingState.Lost,
            Status.LIMITED => TrackingState.Limited,
            Status.EXTENDED_TRACKED => TrackingState.Limited,
            Status.TRACKED => TrackingState.Found,
            _ => throw new ArgumentOutOfRangeException(nameof(state), state, null)
        };
    }
}
```

File: Assets/MirageXR/Player/Scripts/Managers/SpatialMapManager.cs [hunk 4]
```diff
@@ -34,19 +34,26 @@ private void OnDestroy()
...
+                spatialMapMeshs = GameObject.Find("Spatial Awareness System").GetComponentsInChildren<MeshCollider>();
+                //find the spatial map colliders if they are not set
+                //TODO; Find a more efficent way of finding the colliders without using 'GameObject.Find'
+            }
+
+            foreach (MeshCollider spatialMapMesh in spatialMapMeshs)
+            {
+                if (spatialMapMesh)
+                    spatialMapMesh.enabled = !editMode;
+            }
```
Comment (cawke) [inline Assets/MirageXR/Player/Scripts/Managers/SpatialMapManager.cs]:
```suggestion
                if (spatialMapMesh)
                {
                     spatialMapMesh.enabled = !editMode;
                }
            }
```

File: Assets/MirageXR/Player/Scripts/Moodle/Login.cs [hunk 5]
```diff
@@ -341,6 +341,11 @@ public void showLRSSettings()
             ShowPanel(lrsPanel);
         }
 
+        public void showDeveloperOptions()
```
Comment (cawke) [inline Assets/MirageXR/Player/Scripts/Moodle/Login.cs]:
```suggestion
        public void ShowDeveloperOptions()
```

**Review by cawke — CHANGES_REQUESTED** [review]

**Comment by cawke** [comment]
Since you moved LoginMenu to ActivitySelectionMenu, remove CanvasUtility, as this will cause an error.

can i know the reason for moving the prefab? i'm just working on a similar window.

image

**Comment by robhillman97** [comment]
> Since you moved LoginMenu to ActivitySelectionMenu, remove CanvasUtility, as this will cause an error.
> 
> can i know the reason for moving the prefab? i'm just working on a similar window.

We were making this object a child in code in login.cs line 50 with:

`           // make tha login panel a child of the activity list

            transform.position = loginPanelPosition.position;

            transform.rotation = loginPanelPosition.rotation;

            loginCanvas.transform.SetParent(FindObjectOfType<ActivitySelectionMenu>().transform);`

and I wanted to add a reference to the new console window with the login panel without using GameObject.find(); so I added this object to the scene and removed the code in login.cs

File: Assets/MirageXR/Common/Scripts/Managers/ImageTarget/Manager/ImageTargetManagerVuforia.cs [hunk 6]
```diff
@@ -9,19 +8,16 @@ public class ImageTargetManagerVuforia : ImageTargetManagerBase
...
-        {
-            vuforiaRuntime.InitVuforia();
-        }
+        vuforiaRuntime.Initialize();
 
-        while (vuforiaRuntime.InitializationState != VuforiaRuntime.InitState.INITIALIZED)
+        while (!vuforiaRuntime.IsInitialized)
         {
             await Task.Yield();
         }
```
Comment (cawke) [inline Assets/MirageXR/Common/Scripts/Managers/ImageTarget/Manager/ImageTargetManagerVuforia.cs]: add check `if (!vuforiaRuntime.IsInitialized)` before `vuforiaRuntime.Initialize();`

```
        var vuforiaRuntime = VuforiaApplication.Instance;

        if (!vuforiaRuntime.IsInitialized)
        {
            vuforiaRuntime.Initialize();
            while (!vuforiaRuntime.IsInitialized)
            {
                await Task.Yield();
            }
        }
```

**Review by cawke — CHANGES_REQUESTED** [review]

**Review by cawke — APPROVED** [review]

**Pull request merged**